### ISSUE TITLE
File lock on plugin jar not released

### ISSUE THREAD

**Issue by nonnamon** [decebals/pf4j#203]
If a plugin is misconfigured e.g. incorrect Plugin-class name in the manifest file, it is still able to be 'loaded' and 'resolved' but unable to be started or unloaded. 

Is it possible to release the Jar file handle since the plugin cannot be started, such that the file can be over-written? If not then the plugin jar file will be kept open by the process and cannot be overwritten on the file system until process exit.

Conceptually, since the plugin has been 'loaded' it should be possible to 'unload' it too? Currently this fails during attempt to stopPlugin(id)

**Comment by decebals** [comment]
>If a plugin is misconfigured e.g. incorrect Plugin-class name in the manifest file, it is still able to be 'loaded' and 'resolved' 

It's not so clear for me if it's a bug in PF4J or not.
Probably if a plugin class doesn't exist then that plugin should me marked as in error.
All the aspects about the Plugin lifecycle are described in http://www.pf4j.org/doc/plugin-lifecycle.html.

If a plugin is resolved means that all the plugins that depend on the plugin in question can move forward. A plugins uses classes from dependencies and probably this is the reason why the jar is locked.

Do you receive any error? I am curios to know how PF4J signals this issue.

**Comment by nonnamon** [comment]
from memory, a class not found exception happens out of the class loader when attempting to instantiate the class using the incorrect name in the misconfigured manifest. I will try to reproduce with demo and provide some logs.

i agree if the class does not exist should be marked as error, this doesn't seem to be the case currently as the plugin gets to 'resolved'. if a plugin is unable to be used should be able to overwrite the plugin file without restarting the host process.

**Comment by decebals** [comment]
> i agree if the class does not exist should be marked as error, this doesn't seem to be the case currently as the plugin gets to 'resolved'.

Plugin resolved means that jar (storage) exists and the dependencies also and the plugin descriptor is valid (all mandatory fields exist).
If the plugin class is missing from a plugin (jar) then that plugin should be in error and not in resolved state. I think that here is the problem (bug).
I will investigate the case.

**Comment by nonnamon** [comment]
[issue203.txt](https://github.com/decebals/pf4j/files/1799267/issue203.txt)

Logs attached.

e.g. in demo metadat.mf, add required Plugin-xxx fields, use incorrect name 'WelcomePluginX' for Plugin-class. Plugin reaches resolved:

```
2018-03-10 20:33:21,847 DEBUG org.pf4j.AbstractPluginManager - Found descriptor PluginDescriptor [pluginId=welcome-plugin, pluginClass=org.pf4j.demo.welcome.**WelcomePluginX**, version=0.0.1, provider=null, dependencies=[], descr
.....
2018-03-10 20:33:21,862 INFO org.pf4j.AbstractPluginManager - **Plugin 'welcome-plugin@0.0.1' resolved**
2018-03-10 20:33:21,862 TRACE org.pf4j.AbstractPluginManager - Fire 'PluginStateEvent [plugin=welcome-plugin, newState=RESOLVED, oldState=CREATED]' to 'org.pf4j.DefaultExtensionFinder@3b084709'
Plugin 'welcome-plugin':RESOLVED
```

**Comment by HeartUnchange** [comment]
Can you provide the full demo, include the plugin ```welcome-plugin@0.0.1``` and app the use the plugin?

**Comment by nonnamon** [comment]
Attached the demo folder in zip.

Note: the welcome-plugin JAR 

  "demo\app\plugins\pf4j-demo-plugin1-2.3.0-SNAPSHOT.jar"

MANIFEST.MF edited manually to: 

  Plugin-Class: org.pf4j.demo.welcome.WelcomePluginX

**Comment by HeartUnchange** [comment]
I get what you said. 

```PF4J``` may lack a verification for ```loadPlugin```.

**But** why you change ```MANIFEST.MF```  manually?

------------------------------------------------
@decebals  After```pluginClassLoader.addFile(pluginPath.toFile());``` in [JarPluginLoader.java](https://github.com/decebals/pf4j/blob/master/pf4j/src/main/java/org/pf4j/JarPluginLoader.java#L42)
, a verification for the ```Plugin-Class``` maybe better, although I don't think it is necessary.

**Comment by nonnamon** [comment]
> But why you change MANIFEST.MF manually?

The simplest way to reproduce the issue for you was to use the existing PF4J demo projects app & plugin1 with WelcomePlugin, and add a (deliberately misconfigured) MANIFEST.MF

I have been working with JAR / MANIFEST.MF rather than ZIP/plugin.properties, so that is what I'm familiar with. I guess the same could be achieved with plugin.properties.

**Comment by HeartUnchange** [comment]
yes, there is not a proper state for this situation, and adding a verification during loading may be necessary.

@decebals the ```plugin lifestyle``` maybe need to be redesign now.

**Comment by decebals** [comment]
The solution proposed by me is very simple. I hope that is enough.

**Comment by nonnamon** [comment]
It does not seem like a solution to me. 
a) The plugin still reaches RESOLVED state even though it has invalid 'plugin descriptor' 
b) it's presence can prevent proper handling of other valid plugins e.g. if unloadPlugin called
c) the invalid plugin cannot be unloaded and/or it's file overwritten without app restart (on windows)

The solution does not enable being able to unload a improperly configured plugin at runtime without app restart, which is what I intended when I raised the issue.

See attached log.

[pf4j-issue203-with-proposed-fix.zip](https://github.com/decebals/pf4j/files/1831560/pf4j-issue203-with-proposed-fix.zip)
